### ISSUE TITLE
Hashmap iteration is broken for colliding hash code objects

### ISSUE THREAD

**Issue by dankurka** [gwtproject/gwt#9184]
``` java
HashMap<Long, String> testMap = new HashMap<>();
testMap.put(1L | (1L << 32), "one");
testMap.put(1L | (2L << 32), "two");
for (Iterator<Long> it = testMap.keySet().iterator(); it.hasNext();) {
  GWT.log(testMap.get(it.next()));
  it.remove();
}
```

The output should be:

```
one
two
```

But it actually is:

```
one
```

Internal bug number: b/23704249

**Comment by dankurka** [comment]
https://gwt-review.googlesource.com/#/c/13460/

**Comment by niloc132** [comment]
Fixed in 2.8.0-beta1.
